Validate constraints string in Fit

darominski opened the pull request:
**Description of work.**

This PR implements a proposed fix to the found issue, where a hard crash would occur if a Constraints property of the Fit algorithm is filled with an invalid value, specifically if it would not contain an operator (one of "=", "<", ">", etc.). A proposed fix implements a validation of the Constraints property and checks if an operator is present. Otherwise, a standard warning is shown instead of the previously observed hard crash.

**To test:**
1. Run the following script:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import numpy as np

ws = CreateSampleWorkspace()
yAxis = ws.readY(0) 
xAxis = ws.readX(0) 
max_height = np.max(yAxis)
max_index = xAxis[0] + (xAxis[1]-xAxis[0])*(np.where(yAxis == max_height))[0][0]
sigma = 100
fit_fun = "name=Gaussian, PeakCentre={0}, Height={1}, Sigma={2}".format(max_index, 0.7*max_height, sigma)

Fit(
Function=fit_fun,
InputWorkspace=ws,
WorkspaceIndex=1,
Constraints='invalid', # uncomment this line to observe the crash
StartX=float(max_index - 10 * sigma),
EndX=float(max_index + 10 * sigma),
CreateOutput=True,
IgnoreInvalidData=True,
Output='output')
```
2. A warning level message should be displayed informing the user that the provided constraint does not include an operator.
3. There should be no hard crash.

Fixes #34251 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->

Release mention added.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

darominski committed:
Added validation of constraint property string.

The validation currently checks if an operator is present in the string. Lack of thereof was
found to cause a hard crash, while if the operator exists or is incorrect, there exist
validation on the level of parsing the expression in the Expression class.

darominski committed:
Fixed typos in variable names

darominski committed:
Release note for Fit constraint string validation fix

perenon commented:
Hi Dominik, some questions / remarks:
- Would it be feasible to use directly the list of possible operators defined in the API instead of defining `possibleOperators` ?
- Would it be possible to add a test in `Framework/CurveFitting/test/Algorithms/FitTest.h` with a call generating the former crash ?

darominski committed:
Replaced global string list with a const static vector.

darominski committed:
Added unit test for Fig algorithm constraints validation

darominski commented:
> Hi Dominik, some questions / remarks:
> 
>     * Would it be feasible to use directly the list of possible operators defined in the API instead of defining `possibleOperators` ?
> 
>     * Would it be possible to add a test in `Framework/CurveFitting/test/Algorithms/FitTest.h` with a call generating the former crash ?

Hello Remi, thank you for your comments. I completely agree that a unit test would be of great use, and I added it already. You are also right that having a definition of possible operators in two separate places is not a good practice, so I changed the original variable to be a `static const vector` and a member of the `Expression` class rather than a global variable at the namespace level. This way it can be accessed from the `Fit` validation.

darominski committed:
Corrected test to properly test constraints property input

perenon reviewed (dismissed):
Is the code of an acceptable quality?
Does the code conform to the coding standards?
Yes

Are the unit tests small and test the class in isolation?
Yes, the added test in Framework/CurveFitting/test/Algorithms/FitTest is minimal

If there is GUI work does it follow the GUI standards?
No GUI involved

If there are changes in the release notes then do they describe the changes appropriately?
Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?
Changes are put in the bugfixes file

Do changes function as described? Add comments below that describe the tests performed?
Yes. 1- The unittest used to provide a segmentation fault, it does not do this anymore. 2-The app does not crash anymore if the user enters an expression without allowed operator.

Do the changes handle unexpected situations, e.g. bad input?
I did not notice any.

Has the relevant (user and developer) documentation been added/updated?
Yes

martyngigg commented on Framework/API/inc/MantidAPI/Expression.h:
```diff
@@ -116,6 +116,8 @@ class MANTID_API_DLL Expression {
    */
   void toList(const std::string &sep = ",");
 
+  static const std::vector<std::string> DEFAULT_OPS_STR;
```
We do a fair amount of allocations with statics as startup. Would it be possible to switch to using an `std::array<std::string_view>` for this, e.g. https://coliru.stacked-crooked.com/a/a86a8130a624ce70?

perenon replied:
Hello Martyn, sorry we did not reply earlier, we were all on vacation.
I suggest we will discuss that when Dominik is back (September 1st).

darominski replied:
Previously, there was a global variable holding a list of `char *` at the `Mantid::API` namespace level that was subsequently cast to a `std::vector<std::string>`. I replaced the list of `char *` with a `static const std::vector<std::string>`, so there is no need to cast anymore. If there is any speed up possible by replacing `std::string` with `std::string_view`, then it might be lost as either a new `add_operator` function overload needs to be created that accepts `std::string_view` instead of `std::string`, or a cast needs to be performed. Let me know if I am missing something here and whether you would like to pursue this as proposed by you.

martyngigg commented on Framework/CurveFitting/src/Algorithms/Fit.cpp:
```diff
@@ -91,15 +92,37 @@ void Fit::initConcrete() {
+  std::string constraints = getPropertyValue("Constraints");
+  if (constraints.size() > 0) {
+    auto operatorPresent = false;
+    for (auto op : possibleOperators) {
```
```suggestion
    for (const auto & op : possibleOperators) {
```

darominski committed:
Avoid copy of default possible operators

darominski commented on Framework/CurveFitting/src/Algorithms/Fit.cpp:
```diff
@@ -91,15 +92,37 @@ void Fit::initConcrete() {
+std::map<std::string, std::string> Fit::validateInputs() {
+  std::map<std::string, std::string> issues;
+
+  const auto possibleOperators = Mantid::API::Expression::DEFAULT_OPS_STR;
```
I must have missed that, yes, absolutely. Commited.
